write hidden support flag to disk and cache it for longer

AMarcedone opened the pull request.

AMarcedone commented:
Busting the cache on any error (including i.e. network errors) is a bit aggressive, but I do not want to  risk forgetting some specific error type one way or the other. Open to suggestions.

Also this is hard to test because the feature flag is meant to only work in prod. I plan to do some manual testing against prod with a puppet account. Other alternatives are possible but more time consuming.

AMarcedone commented:
Performing the review a commit at a time might help.

maxtaco commented on go/teams/loader.go:
```diff
@@ -469,6 +476,13 @@ func (l *TeamLoader) load2(ctx context.Context, arg load2ArgT) (ret *load2ResT,
 	defer l.G().CTraceTimed(ctx, traceLabel, func() error { return err })()
+	defer func() {
+		if err != nil {
+			mctx := libkb.NewMetaContext(ctx, l.G())
```
make this metacontext when entering the function, after log tags are assigned.

maxtaco replied:
Also, simplify, just do this check after the call to `load2Inner` fails, i.e., no need for a `defer`

AMarcedone commented on go/teams/audit.go:
```diff
@@ -146,6 +146,12 @@ func (a *Auditor) AuditTeam(m libkb.MetaContext, id keybase1.TeamID, isPublic bo
 			Ctime:     keybase1.ToTime(start),
 		})
 	}()
+	defer func() {
```
@maxtaco do you want me to change this one too? The defer above this one can also be simplified, so not sure what the best thing to do is....

maxtaco replied:
urg, some of the ugliness is pre-existing here. In general, I'd say two use a `defer` if you'd otherwise need to insert that code at `>1` callsites. Which as you point out isn't the case here.

heronhaye commented on go/teams/hidden/manager.go:
```diff
@@ -40,19 +35,25 @@ type loadArg struct {
+	// if we never checked before or the chain was not supported but the cache
+	// expired, check again. Once enabled, hidden support cannot be revoked
+	// regardless of the cache staleness.
+	if supportsHiddenState == nil || (!supportsHiddenState.State && mctx.G().Clock().Now().Before(supportsHiddenState.CacheUntil)) {
```
>mctx.G().Clock().Now().Before(supportsHiddenState.CacheUntil)

is this inverted?

AMarcedone replied:
good catch! thanks!

heronhaye commented on go/teams/storage/supports_hidden_flag.go:
```diff
@@ -0,0 +1,90 @@
+
+// Increment to invalidate the disk cache.
+const supportsHiddenFlagDiskStorageVersion = 1
+const supportsHiddenFlagCacheLRUSize = 200
```
Could make it bigger. For bots

AMarcedone replied:
What do you think is a good value here? 1k? 10k?

heronhaye commented on go/teams/storage/supports_hidden_flag.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+func NewSupportsHiddenFlagStorage(g *libkb.GlobalContext) *SupportsHiddenFlagStorage {
+	s := newStorageGeneric(g, supportsHiddenFlagCacheLRUSize, supportsHiddenFlagDiskStorageVersion, libkb.DBSupportsHiddenFlagStorage, libkb.EncryptionReasonTeamsHiddenLocalStorage, "hidden flag", func() diskItemGeneric { return &supportsHiddenFlagDiskStorageItem{} })
```
try to keep <100 chars/line

heronhaye commented on go/teams/storage/supports_hidden_flag.go:
```diff
@@ -0,0 +1,90 @@
+	}
+	ret, ok := vp.(*HiddenChainSupportState)
+	if !ok {
+		mctx.Debug("teams.storage/SupportsHiddenFlagStorage#Get cast error: %T is wrong type", vp)
```
should it return nil in this case? (it might do that anyway, not sure)

AMarcedone replied:
It does: https://tour.golang.org/methods/15 . I am adding a return nil for extra clarity

heronhaye commented on go/teams/storage/supports_hidden_flag.go:
```diff
@@ -0,0 +1,90 @@
+func (s *SupportsHiddenFlagStorage) ClearEntryIfFalse(mctx libkb.MetaContext, teamID keybase1.TeamID) {
+	if currentState := s.Get(mctx, teamID); currentState != nil && !currentState.State {
+		// put an already expired state in the cache
+		s.Put(mctx, &HiddenChainSupportState{TeamID: teamID, State: false, CacheUntil: mctx.G().Clock().Now().Add(-1 * time.Second)})
```
sure, just curious why this instead of putting nil?

AMarcedone replied:
It was preexisting, and I kept it since I thought it would be more clear. 
I am changing it.

AMarcedone commented:
@heronhaye did some extra manual testing since you approved. As a result I added more logging, and prevented APINetError from busting the cache (i.e. if your phone goes offline). 
I also changed the nil value to a struct with the proper teamID but empty CacheUntil because teamID is necessary to prevent crashes.

Feel free to take another look, I will merge unless I hear otherwise.
